internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -192,42 +215,186 @@ func (r *NumaflowControllerRolloutReconciler) reconcile(
+				numaLogger.Info("Numaflow Controller either needs to or is in the process of updating")
+
+				// request pause if we haven't already
+				pauseRequestUpdated, err := r.requestPipelinesPause(ctx, controllerRollout, true)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+				if !pauseRequestUpdated { // don't do this yet if we just made a request - it's too soon for anything to have happened
+					// check if the pipelines are all paused and we're trying to update the spec
+					if controllerDeploymentNeedsUpdating {
+						allPaused, err := r.allPipelinesPaused(ctx, namespace)
+						if err != nil {
+							return ctrl.Result{}, err
+						}
+						if allPaused {
+							numaLogger.Infof("confirmed all Pipelines have paused so Numaflow Controller can safely update")
+							// apply controller
+							phase, err := r.sync(controllerRollout, namespace, numaLogger)
+							if err != nil {
+								return ctrl.Result{}, err
+							}
+							if phase != gitopsSyncCommon.OperationSucceeded {
+								return ctrl.Result{}, fmt.Errorf("sync operation is not successful")
+							}
+						} else {
+							numaLogger.Debugf("not all Pipelines have paused")
+						}
+					}
+				}
+				return delayedRequeue, nil
+			} else {
+				// remove any pause requirement if necessary
+				_, err := r.requestPipelinesPause(ctx, controllerRollout, false)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+		}
+
+	}
+
+	// apply controller - this handles syncing in the cases in which our Controller Rollout isn't updating
+	// (note that the cases above in which it is updating have a 'return' statement):
+	// - new ControllerRollout
+	// - auto healing
+	// - somebody changed the manifest associated with the Controller version (shouldn't happen but could)
 	phase, err := r.sync(controllerRollout, namespace, numaLogger)
 	if err != nil {
-		return err
+		return ctrl.Result{}, err
 	}
 
 	if phase != gitopsSyncCommon.OperationSucceeded {
-		return fmt.Errorf("sync operation is not successful")
+		return ctrl.Result{}, fmt.Errorf("sync operation is not successful")
 	}
 
-	err = r.processNumaflowControllerStatus(ctx, controllerRollout)
+	return ctrl.Result{}, nil
+}
+
+func (r *NumaflowControllerRolloutReconciler) getPipelines(ctx context.Context, namespace string) ([]*kubernetes.GenericObject, error) {
+	return kubernetes.ListCR(ctx, r.restConfig, common.NumaflowAPIGroup, common.NumaflowAPIVersion, "pipelines", namespace, "", "")
+}
+
+func (r *NumaflowControllerRolloutReconciler) allPipelinesPaused(ctx context.Context, namespace string) (bool, error) {
+	pipelines, err := r.getPipelines(ctx, namespace)
 	if err != nil {
-		return err
+		return false, err
 	}
+	for _, pipeline := range pipelines {
+		status, err := kubernetes.ParseStatus(pipeline)
+		if err != nil {
+			return false, err
+		}
+		if status.Phase != "Paused" {
+			return false, nil
+		}
+	}
+	return true, nil
+}
```

Comment:
Yes, maybe if both Reconcilers implement a common interface which includes a `getPipelines()` function, then this could a single function that uses it.